---
title: "HW 09: Strings"
author: "Madison Hongsermeier-Baxter"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---

```{r}
library(tidyverse)
install.packages("ggthemes")
library(ggthemes)
install.packages("RColorBrewer")
install.packages("viridis")
```

# Part 1: DNA strings

## Import Data

### Import beta_gliadin.txt using the read_lines() function. Skip the first line.
```{r}
beta_gliadin <- read_lines("~/Documents/BI485/hongsermeier-baxter_madison/data/beta_gliadin.txt", skip = 1)
beta_gliadin
```

## From many strings to one

### Remove unwanted characters.
```{r}
beta_gliadin1 <- beta_gliadin %>% 
  str_replace_all("\\s", "\\") %>% 
  str_replace_all("\\d", "\\") %>% 
  str_to_upper()
beta_gliadin1
```

### Collapse strings.
```{r}
beta_gliadin2 <- beta_gliadin %>% 
  str_replace_all("\\s", "\\") %>% 
  str_replace_all("\\d", "\\") %>% 
  str_c(collapse = "") %>% 
  str_to_upper()
beta_gliadin2
```

### str_length() returns the length of a string. How long is the total DNA sequence? You should get 3310 characters.
```{r}
str_length(beta_gliadin2)
```

## Sequence analysis: find the promoter region

### Use str_count() and regex to count the number of IUPAC letters in the sequence that are not A, C, G, or T.
```{r}
str_count(beta_gliadin2, c("A", "T", "C", "G"))
```

#### A, C, G, and T account for all of the letters

### Count how many possible start codons are in the sequence. Count how many possible stop codons are located in the sequence.
```{r}
str_count(beta_gliadin2, "ATG")
str_count(beta_gliadin2, "TAA|TGA")
```

### Assign the two motif sequences to variables caat_box and tata_box, respectively.
```{r}
caat_box <- "CCAAT"
tata_box <- "CTATAA"
```

### Count how many times each of these motifs occurs in the DNA sequence.
```{r}
str_count(beta_gliadin2, caat_box)
str_count(beta_gliadin2, tata_box)
```

### Make a regex search string called caat_seq that begins with “CCAAT”, ends with “ATG”", and allows for a stretch of 100-150 nucleotides between the two. Count how many times this possible sequence occurs.
```{r}
caat_seq <- "CCAAT.{100,150}ATG"
str_count(beta_gliadin2, caat_seq)
```

###  Make a regex search string called tata_seq that begins with “CTATAA”, ends with “ATG”, and allows for a stretch of 50-100 nucleotides between the two. Count how many times this possible sequence occurs.
```{r}
tata_seq <- "CTATAA.{50,100}ATG"
str_count(beta_gliadin2, tata_seq)
```

### Use str_extract_all to extract all of the caat_seq sequences and save them in a variable called caat_seqs_all.
```{r}
caat_seqs_all <- str_extract_all(beta_gliadin2, caat_seq, simplify = TRUE)
str_count(caat_seqs_all)
```

### Use str_which() to find which caat_seq has the tata_box sequence. Use that value to extract the full promotor sequence into a variable called promotor. Print promotor and determine it’s length.

```{r}
str_which(caat_seqs_all, tata_box)
promoter <- str_subset(caat_seqs_all, tata_box)
promoter
str_count(promoter)
```

# Part 2: Ants

## Import
```{r}
ants <- read.csv("~/Documents/BI485/hongsermeier-baxter_madison/data/ants.csv")
ants
```

## Wrangle 

### Tidy data.
```{r}
library(stringr)
ants1 <- ants %>% 
  select(-c(season, seas.code, trt, plot, sample)) %>% 
  gather(key = species, value = abundance, matches("\\w\\.\\w", ignore.case = FALSE))
ants1
```

### Reduce the long species names to eight letter CEP names. Use mutate to turn the year into an order factor with levels “2010” and “2011”. Then, group by year and species, and summarize the total abundance of each species.
```{r}
ants2 <- ants1 %>% 
  mutate(
    species = str_replace(
      species,
      pattern = "(^[A-Z]*\\w{3})\\w{3,}\\.(\\w{4})\\w{2,}",
      replacement = "\\1\\2"),
    year = factor(year, levels = c("2010", "2011"), labels = c("2010", "2011"), ordered = TRUE)) %>% 
  group_by(year, species) %>% 
  summarise(total = sum(abundance, na.rm = TRUE)
  )
ants2
```

## Plot the data.
```{r}
ggplot(data = ants2, aes(x = reorder(species, total), y = total, group = species)) + 
  geom_line(color = "gray") +
  geom_point( aes(shape = year), size = 2) +
  coord_flip() +
  labs(y = "Total Abundance", x = NULL, shape = "Year") +
  theme_minimal()
```

# Part 3:Phlox Flowers

## Import and wrangle first data set

### Import.
```{r}
flower_size <- read.csv("~/Documents/BI485/hongsermeier-baxter_madison/data/flower_size.csv")
flower_size
```

### Filter out rows where Flower number is not NA.
```{r}
flower_size1 <- flower_size %>% 
  filter(!is.na(Flower.number))
flower_size1
```

### Select the “Species”, “Corolla length (cm)”, and “Corolla width throat (cm)” columns. Make the column names syntactic.
```{r}
flower_size2 <- flower_size1 %>% 
  select(species = Species, cor_length = Corolla.length..cm., throat_width = Corolla.width.throat..cm.)
flower_size2
```

### Use separate to separate the species column into genus and species columns, dropping everything else.
```{r}
flower_size3 <- flower_size2 %>% 
  separate(species, into = c("genus", "species"))
 flower_size3
```

### Group the data by genus and species to summarize the mean() and max() lengths and widths for each species.
```{r}
flower_size4 <- flower_size3 %>% 
  group_by(genus, species) %>% 
  summarize(mean_length = mean(cor_length), mean_width = mean(throat_width), max_length = max(cor_length), max_width = max(throat_width))
flower_size4
```

## Import and wrangle second data set.

### Import.
```{r}
flower_pollinators <- read.csv("~/Documents/BI485/hongsermeier-baxter_madison/data/flower_pollinators.csv", na.strings = c("", "NA"))
flower_pollinators
```

### Use select() to remove columns that start with “Source” and to remove the Color column.
```{r}
flower_pollinators1 <- flower_pollinators %>% 
  select(-c(Color, starts_with("Source")))
flower_pollinators1
```

### Use filter() to remove rows where Pollinator is NA.
```{r}
flower_pollinators2 <- flower_pollinators1 %>%
  filter(!is.na(Pollinator))
flower_pollinators2
```

### Use separate() as you did above to separate Species into genus and species columns. Again, I used lower case letters for genus and species.
```{r}
flower_pollinators2 %>%
  separate(Species, into = c("genus", "species"), sep = " ", extra = "drop")
```

### Save these last three steps in a variable called pollinators.
```{r}
pollinators <- flower_pollinators2 %>%
  separate(Species, into = c("genus", "species"), sep = " ", extra = "drop")
pollinators
```

### Use separate() to separate the first pollinator from all other words. The sep argument must be a regex pattern that finds either a comma or a space. Save the result in a column called Syndrome.
```{r}
pollinators <- pollinators %>% 
    separate(Pollinator, into = "Syndrome", sep = " |, ", extra = "drop")
pollinators
```

## Join the datasets

### Use inner_join to join the smaller data set to the larger set. As long as your column names for genus and species are identical in both tibbles (e.g., genus and species), the join function will automatically match both columns when merging the data.
```{r}
phlox <- inner_join(flower_size4, pollinators, by = c("genus", "species"))
phlox
```

## Plotting

### Make a scatterplot to show the relationship between mean_length (x-axis) and mean_width (y-axis). Color the points by Syndrome.
```{r}

```

### Use mutate with str_replace() and regex patterns to replace all occurences of “beetle” with “beetles,” and “bee” with “bees”. Be careful. If you don’t do this right, then you could end up with “beetless”. Make use of the $ match character.
```{r}

```

### Replot your results once you are sure you replaced the names correctly.
```{r}

```

### Once again, use mutate() with str_replace and a regex pattern to replace all instances of “beetles” or “flies” with “other insects”.
```{r}

```

### Plot the results again.
```{r}

```

### Use mutate and case_when to create a new column called syndrome_group and to change autogamous to Autogamous (upper case A), bat to Mammal, hummingbird to Bird, and everything else to Insect. Creating a new column would allow you to access the individual syndromes for future plots, when necessary.
```{r}

```

### Replot the results.
```{r}

```












