---
title: "HW 09: Strings"
author: "Madison Hongsermeier-Baxter"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---

```{r}
library(tidyverse)
install.packages("ggthemes")
library(ggthemes)
install.packages("RColorBrewer")
install.packages("viridis")
```

# Part 1: DNA strings

## Import Data

### Import beta_gliadin.txt using the read_lines() function. Skip the first line.
```{r}
beta_gliadin <- read_lines("~/Documents/BI485/hongsermeier-baxter_madison/data/beta_gliadin.txt", skip = 1)
beta_gliadin
```

## From many strings to one

### Remove unwanted characters.
```{r}
beta_gliadin1 <- beta_gliadin %>% 
  str_replace_all("\\s", "\\") %>% 
  str_replace_all("\\d", "\\") %>% 
  str_to_upper()
beta_gliadin1
```

### Collapse strings.
```{r}
beta_gliadin2 <- beta_gliadin %>% 
  str_replace_all("\\s", "\\") %>% 
  str_replace_all("\\d", "\\") %>% 
  str_c(collapse = "") %>% 
  str_to_upper()
beta_gliadin2
```

### str_length() returns the length of a string. How long is the total DNA sequence? You should get 3310 characters.
```{r}
str_length(beta_gliadin2)
```

## Sequence analysis: find the promoter region

### Use str_count() and regex to count the number of IUPAC letters in the sequence that are not A, C, G, or T.
```{r}
str_count(beta_gliadin2, c("A", "T", "C", "G"))
```

#### A, C, G, and T account for all of the letters

### Count how many possible start codons are in the sequence. Count how many possible stop codons are located in the sequence.
```{r}
str_count(beta_gliadin2, "ATG")
str_count(beta_gliadin2, "TAA|TGA")
```

### Assign the two motif sequences to variables caat_box and tata_box, respectively.
```{r}
caat_box <- "CCAAT"
tata_box <- "CTATAA"
```

### Count how many times each of these motifs occurs in the DNA sequence.
```{r}
str_count(beta_gliadin2, caat_box)
str_count(beta_gliadin2, tata_box)
```

### Make a regex search string called caat_seq that begins with “CCAAT”, ends with “ATG”", and allows for a stretch of 100-150 nucleotides between the two. Count how many times this possible sequence occurs.
```{r}
caat_seq <- "CCAAT.{100,150}ATG"
str_count(beta_gliadin2, caat_seq)
```

###  Make a regex search string called tata_seq that begins with “CTATAA”, ends with “ATG”, and allows for a stretch of 50-100 nucleotides between the two. Count how many times this possible sequence occurs.
```{r}
tata_seq <- "CTATAA.{50,100}ATG"
str_count(beta_gliadin2, tata_seq)
```

### Use str_extract_all to extract all of the caat_seq sequences and save them in a variable called caat_seqs_all.
```{r}
caat_seqs_all <- str_extract_all(beta_gliadin2, caat_seq, simplify = TRUE)
str_count(caat_seqs_all)
```

### Use str_which() to find which caat_seq has the tata_box sequence. Use that value to extract the full promotor sequence into a variable called promotor. Print promotor and determine it’s length.

```{r}
str_which(caat_seqs_all, tata_box)
promoter <- str_subset(caat_seqs_all, tata_box)
promoter
str_count(promoter)
```



    

  